File: Raven.Client.Lightweight/Linq/RavenQueryInspector.cs
//-----------------------------------------------------------------------
// <copyright file="RavenQueryInspector.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Raven.Abstractions.Data;
using Raven.Abstractions.Extensions;
using Raven.Abstractions.Indexing;
using Raven.Client.Connection;
using Raven.Client.Document;
using Raven.Client.Connection.Async;
using Raven.Client.Indexes;
using Raven.Client.Spatial;
using Raven.Json.Linq;

namespace Raven.Client.Linq
{
	/// <summary>
	/// Implements <see cref="IRavenQueryable{T}"/>
	/// </summary>
	public class RavenQueryInspector<T> : IRavenQueryable<T>, IRavenQueryInspector
	{
		private readonly Expression expression;
		private readonly IRavenQueryProvider provider;
		private readonly RavenQueryStatistics queryStats;
		private readonly RavenQueryHighlightings highlightings;
		private readonly string indexName;
		private readonly IDatabaseCommands databaseCommands;
		private readonly IAsyncDatabaseCommands asyncDatabaseCommands;
		private InMemoryDocumentSessionOperations session;
		private readonly bool isMapReduce;

		/// <summary>
		/// Initializes a new instance of the <see cref="RavenQueryInspector{T}"/> class.
		/// </summary>
		public RavenQueryInspector(
			IRavenQueryProvider provider, 
			RavenQueryStatistics queryStats,
			RavenQueryHighlightings highlightings,
			string indexName,
			Expression expression,
			InMemoryDocumentSessionOperations session
			, IDatabaseCommands databaseCommands
			, IAsyncDatabaseCommands asyncDatabaseCommands,
			bool isMapReduce
			)
		{
			if (provider == null)
			{
				throw new ArgumentNullException("provider");
			}
			this.provider = provider.For<T>();
			this.queryStats = queryStats;
			this.highlightings = highlightings;
			this.indexName = indexName;
			this.session = session;
			this.databaseCommands = databaseCommands;
			this.asyncDatabaseCommands = asyncDatabaseCommands;
			this.isMapReduce = isMapReduce;
			this.provider.AfterQueryExecuted(this.AfterQueryExecuted);
			this.expression = expression ?? Expression.Constant(this);
		}

		private void AfterQueryExecuted(QueryResult queryResult)
		{
			this.queryStats.UpdateQueryStats(queryResult);
			this.highlightings.Update(queryResult);
		}

		#region IOrderedQueryable<T> Members

		Expression IQueryable.Expression
		{
			get { return expression; }
		}

		Type IQueryable.ElementType
		{
			get { return typeof(T); }
		}

		IQueryProvider IQueryable.Provider
		{
			get { return provider; }
		}

		/// <summary>
		/// Gets the enumerator.
		/// </summary>
		/// <returns></returns>
		public IEnumerator<T> GetEnumerator()
		{
			var execute = provider.Execute(expression);
			return ((IEnumerable<T>)execute).GetEnumerator();
		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		#endregion

		/// <summary>
		/// Provide statistics about the query, such as total count of matching records
		/// </summary>
		public IRavenQueryable<T> Statistics(out RavenQueryStatistics stats)
		{
			stats = queryStats;
			return this;
		}

		/// <summary>
		/// Customizes the query using the specified action
		/// </summary>
		/// <param name="action">The action.</param>
		/// <returns></returns>
		public IRavenQueryable<T> Customize(Action<IDocumentQueryCustomization> action)
		{
			provider.Customize(action);
			return this;
		}

	    public IRavenQueryable<TResult> TransformWith<TTransformer, TResult>() where TTransformer : AbstractTransformerCreationTask, new()
	    {
	        var transformer = new TTransformer();
	        provider.TransformWith(transformer.TransformerName);
	        return (IRavenQueryable<TResult>)this.As<TResult>();
	    }

        public IRavenQueryable<TResult> TransformWith<TResult>(string transformerName)
        {
            provider.TransformWith(transformerName);
            return (IRavenQueryable<TResult>)this.As<TResult>();
        }

	    public IRavenQueryable<T> AddQueryInput(string input, RavenJToken foo)
	    {
	        provider.AddQueryInput(input, foo);
	        return this;
	    }

		public IRavenQueryable<T> Spatial(Expression<Func<T, object>> path, Func<SpatialCriteriaFactory, SpatialCriteria> clause)
		{
			return Customize(x => x.Spatial(path.ToPropertyPath(), clause));
		}

		/// <summary>
		/// Returns a <see cref="System.String"/> that represents this instance.
		/// </summary>
		/// <returns>
		/// A <see cref="System.String"/> that represents this instance.
		/// </returns>
		public override string ToString()
		{
			RavenQueryProviderProcessor<T> ravenQueryProvider = GetRavenQueryProvider();
			var documentQuery = ravenQueryProvider.GetDocumentQueryFor(expression);
			string fields = "";
			if (ravenQueryProvider.FieldsToFetch.Count > 0)
				fields = "<" + string.Join(", ", ravenQueryProvider.FieldsToFetch.ToArray()) + ">: ";
			return fields + documentQuery;
		}

		public IndexQuery GetIndexQuery(bool isAsync = true)
		{
			RavenQueryProviderProcessor<T> ravenQueryProvider = GetRavenQueryProvider();
			if (isAsync == false)
			{
				var documentQuery = ravenQueryProvider.GetDocumentQueryFor(expression);
				return documentQuery.GetIndexQuery(false);
			}
			var asyncDocumentQuery = ravenQueryProvider.GetAsyncDocumentQueryFor(expression);
			return asyncDocumentQuery.GetIndexQuery(true);
		}

		public virtual FacetResults GetFacets(string facetSetupDoc, int start, int? pageSize)
		{
			return databaseCommands.GetFacets(indexName, GetIndexQuery(false), facetSetupDoc, start, pageSize);
		}

		public virtual FacetResults GetFacets(List<Facet> facets, int start, int? pageSize)
		{
			return databaseCommands.GetFacets(indexName, GetIndexQuery(false), facets, start, pageSize);
		}

		public virtual Task<FacetResults> GetFacetsAsync(string facetSetupDoc, int start, int? pageSize)
		{
			return asyncDatabaseCommands.GetFacetsAsync(indexName, GetIndexQuery(true), facetSetupDoc, start, pageSize);
		}

		public virtual Task<FacetResults> GetFacetsAsync(List<Facet> facets, int start, int? pageSize)
		{
			return asyncDatabaseCommands.GetFacetsAsync(indexName, GetIndexQuery(true), facets, start, pageSize);
		}

		private RavenQueryProviderProcessor<T> GetRavenQueryProvider()
		{
		    return new RavenQueryProviderProcessor<T>(provider.QueryGenerator, provider.CustomizeQuery, null, indexName,
		                                              new HashSet<string>(), new List<RenamedField>(), isMapReduce,
                                                      provider.ResultTransformer, provider.QueryInputs);
		}

		/// <summary>
		/// Get the name of the index being queried
		/// </summary>
		public string IndexQueried
		{
			get
			{
				var ravenQueryProvider = new RavenQueryProviderProcessor<T>(provider.QueryGenerator, null, null, indexName, new HashSet<string>(), new List<RenamedField>(), isMapReduce, 
                    provider.ResultTransformer, provider.QueryInputs);
				var documentQuery = ravenQueryProvider.GetDocumentQueryFor(expression);
				return ((IRavenQueryInspector)documentQuery).IndexQueried;
			}
		}

		/// <summary>
		/// Get the name of the index being queried asynchronously
		/// </summary>
		public string AsyncIndexQueried
		{
			get
			{
				var ravenQueryProvider = new RavenQueryProviderProcessor<T>(provider.QueryGenerator, null, null, indexName, new HashSet<string>(), new List<RenamedField>(), isMapReduce,
                    provider.ResultTransformer, provider.QueryInputs);
				var documentQuery = ravenQueryProvider.GetAsyncDocumentQueryFor(expression);
				return ((IRavenQueryInspector)documentQuery).IndexQueried;
			}
		}

		/// <summary>
		/// Grant access to the database commands
		/// </summary>
		public IDatabaseCommands DatabaseCommands
		{
			get
			{
				if(databaseCommands == null)
					throw new NotSupportedException("You cannot get database commands for this query");
				return databaseCommands;
			}
		}

		/// <summary>
		/// Grant access to the async database commands
		/// </summary>
		public IAsyncDatabaseCommands AsyncDatabaseCommands
		{
			get
			{
				if (asyncDatabaseCommands == null)
					throw new NotSupportedException("You cannot get database commands for this query");
				return asyncDatabaseCommands;
			}
		}

		public InMemoryDocumentSessionOperations Session
		{
			get
			{
				return session;
			}
		}

		///<summary>
		/// Get the last equality term for the query
		///</summary>
		public KeyValuePair<string, string> GetLastEqualityTerm(bool isAsync = false)
		{
            var ravenQueryProvider = new RavenQueryProviderProcessor<T>(provider.QueryGenerator, null, null, indexName, new HashSet<string>(), new List<RenamedField>(), isMapReduce, provider.ResultTransformer, provider.QueryInputs);
			if (isAsync)
			{
				var asyncDocumentQuery = ravenQueryProvider.GetAsyncDocumentQueryFor(expression);
				return ((IRavenQueryInspector)asyncDocumentQuery).GetLastEqualityTerm(true);
			}

			var documentQuery = ravenQueryProvider.GetDocumentQueryFor(expression);
			return ((IRavenQueryInspector) documentQuery).GetLastEqualityTerm();
		}

		/// <summary>
		/// Set the fields to fetch
		/// </summary>
		public void FieldsToFetch(IEnumerable<string> fields)
		{
			foreach (var field in fields)
			{
				provider.FieldsToFetch.Add(field);
			}
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/client/linq/RavenQueryInspector.java
package raven.client.linq;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

import org.apache.commons.lang.StringUtils;

import com.mysema.query.support.Expressions;
import com.mysema.query.types.Expression;
import com.mysema.query.types.OrderSpecifier;
import com.mysema.query.types.Path;
import com.mysema.query.types.Predicate;

import raven.abstractions.basic.Lazy;
import raven.abstractions.basic.Reference;
import raven.abstractions.basic.Tuple;
import raven.abstractions.closure.Action1;
import raven.abstractions.closure.Function1;
import raven.abstractions.data.Facet;
import raven.abstractions.data.FacetResults;
import raven.abstractions.data.IndexQuery;
import raven.abstractions.data.QueryResult;
import raven.abstractions.extensions.ExpressionExtensions;
import raven.abstractions.json.linq.RavenJToken;
import raven.client.IDocumentQuery;
import raven.client.IDocumentQueryCustomization;
import raven.client.RavenQueryHighlightings;
import raven.client.RavenQueryStatistics;
import raven.client.connection.IDatabaseCommands;
import raven.client.connection.IRavenQueryInspector;
import raven.client.document.InMemoryDocumentSessionOperations;
import raven.client.indexes.AbstractTransformerCreationTask;
import raven.client.spatial.SpatialCriteria;
import raven.client.spatial.SpatialCriteriaFactory;
import raven.linq.dsl.LinqOps;

public class RavenQueryInspector<T> implements IRavenQueryable<T>, IRavenQueryInspector {

  private Class<T> clazz;
  private final Expression<?> expression;
  private IRavenQueryProvider provider;
  private final RavenQueryStatistics queryStats;
  private final RavenQueryHighlightings highlightings;
  private final String indexName;
  private final IDatabaseCommands databaseCommands;
  private InMemoryDocumentSessionOperations session;
  private final boolean isMapReduce;

  public RavenQueryInspector(Class<T> clazz, IRavenQueryProvider provider, RavenQueryStatistics queryStats, RavenQueryHighlightings highlightings, String indexName, Expression<?> expression, InMemoryDocumentSessionOperations session, IDatabaseCommands databaseCommands, boolean isMapReduce) {
    this.clazz = clazz;
    if (provider == null) {
      throw new IllegalArgumentException("provider is null");
    }
    this.provider = provider.forClass(clazz);
    this.queryStats = queryStats;
    this.highlightings = highlightings;
    this.indexName = indexName;
    this.session = session;
    this.databaseCommands = databaseCommands;
    this.isMapReduce = isMapReduce;
    this.provider.afterQueryExecuted(new Action1<QueryResult>() {

      @Override
      public void apply(QueryResult queryResult) {
        afterQueryExecuted(queryResult);
      }
    });
    this.expression = expression != null ? expression : Expressions.constant("root");
  }

  private void afterQueryExecuted(QueryResult queryResult) {
    this.queryStats.updateQueryStats(queryResult);
    this.highlightings.update(queryResult);
  }

  @Override
  public Expression< ? > getExpression() {
    return expression;
  }

  @Override
  public Class< ? > getElementType() {
    return clazz;
  }

  @Override
  public IQueryProvider getProvider() {
    return provider;
  }

  @Override
  public IRavenQueryable<T> where(Predicate predicate) {
    return getProvider().createQuery(Expressions.operation(expression.getType(), LinqOps.Query.WHERE, expression, predicate));
  }

  @SuppressWarnings("unchecked")
  @Override
  public List<T> toList() {
    Object execute = getProvider().execute(expression);
    return ((IDocumentQuery<T>)execute).toList();
  }

  @SuppressWarnings("unchecked")
  @Override
  public T single() {
    Object execute = getProvider().execute(expression);
    List<T> list = ((IDocumentQuery<T>)execute).toList();
    if (list.size() != 1) {
      throw new IllegalStateException("Expected single result! Got: " + list.size());
    }
    return list.get(0);
  }

  /**
   * Provide statistics about the query, such as total count of matching records
   */
  @Override
  public IRavenQueryable<T> statistics(Reference<RavenQueryStatistics> stats) {
    stats.value = queryStats;
    return this;
  }

  /**
   * Customizes the query using the specified action
   */
  @Override
  public IRavenQueryable<T> customize(Action1<IDocumentQueryCustomization> action) {
    provider.customize(action);
    return this;
  }

  @Override
  public <S> IRavenQueryable<S> transformWith(Class< ? extends AbstractTransformerCreationTask> transformerClazz, Class<S> resultClass) {
    try {
      AbstractTransformerCreationTask transformer = transformerClazz.newInstance();
      provider.transformWith(transformer.getTransformerName());
      return (IRavenQueryable<S>) as(resultClass);
    } catch (Exception e){
      throw new RuntimeException(e);
    }
  }


  @Override
  public IRavenQueryable<T> addQueryInput(String name, RavenJToken value) {
    provider.addQueryInput(name, value);
    return this;
  }

  @Override
  public IRavenQueryable<T> spatial(final Path< ? > path, final Function1<SpatialCriteriaFactory, SpatialCriteria> clause) {
    return customize(new Action1<IDocumentQueryCustomization>() {

      @Override
      public void apply(IDocumentQueryCustomization x) {
        x.spatial(ExpressionExtensions.toPropertyPath(path), clause);
      }
    });
  }

  @Override
  public String toString() {
    RavenQueryProviderProcessor<T> ravenQueryProvider = getRavenQueryProvider();
    IDocumentQuery<T> luceneQuery = ravenQueryProvider.getLuceneQueryFor(expression);
    String fields = "";
    if (!ravenQueryProvider.getFieldsToFetch().isEmpty()) {
      fields = "<" + StringUtils.join(ravenQueryProvider.getFieldsToFetch(), ", ") + ">: ";
    }
    return fields + luceneQuery;
  }

  @Override
  public IndexQuery getIndexQuery() {
    RavenQueryProviderProcessor<T> ravenQueryProvider = getRavenQueryProvider();
    IDocumentQuery<T> luceneQuery = ravenQueryProvider.getLuceneQueryFor(expression);
    return luceneQuery.getIndexQuery();
  }

  @Override
  public FacetResults getFacets(String facetSetupDoc, int start, Integer pageSize) {
    return databaseCommands.getFacets(indexName, getIndexQuery(), facetSetupDoc, start, pageSize);
  }

  @Override
  public FacetResults getFacets(List<Facet> facets, int start, Integer pageSize) {
    return databaseCommands.getFacets(indexName, getIndexQuery(), facets, start, pageSize);
  }

  private RavenQueryProviderProcessor<T> getRavenQueryProvider() {
    return new RavenQueryProviderProcessor<>(clazz, provider.getQueryGenerator(), provider.getCustomizeQuery(), null, indexName,
        new HashSet<String>(), new ArrayList<RenamedField>(), isMapReduce, provider.getResultTranformer(), provider.getQueryInputs());
  }

  @Override
  public String getIndexQueried() {
    RavenQueryProviderProcessor<T> ravenQueryProvider = new RavenQueryProviderProcessor<>(clazz, provider.getQueryGenerator(), null, null, indexName, new HashSet<String>(),
        new ArrayList<RenamedField>(), isMapReduce, provider.getResultTranformer(), provider.getQueryInputs());
    IDocumentQuery<T> luceneQuery = ravenQueryProvider.getLuceneQueryFor(expression);
    return ((IRavenQueryInspector)luceneQuery).getIndexQueried();
  }

  @Override
  public IDatabaseCommands getDatabaseCommands() {
    if (databaseCommands == null) {
      throw new IllegalStateException("You cannot get database commands for this query");
    }
    return databaseCommands;
  }

  @Override
  public InMemoryDocumentSessionOperations getSession() {
    return session;
  }

  @Override
  public Tuple<String, String> getLastEqualityTerm() {
    RavenQueryProviderProcessor<T> ravenQueryProvider = new RavenQueryProviderProcessor<>(clazz, provider.getQueryGenerator(), null, null, indexName, new HashSet<String>(),
        new ArrayList<RenamedField>(), isMapReduce, provider.getResultTranformer(), provider.getQueryInputs());
    IDocumentQuery<T> luceneQuery = ravenQueryProvider.getLuceneQueryFor(expression);
    return ((IRavenQueryInspector) luceneQuery).getLastEqualityTerm();
  }

  public void fieldsToFetch(List<String> fields) {
    for (String field: fields) {
      provider.getFieldsToFetch().add(field);
    }
  }

  @Override
  public Lazy<List<T>> lazily() {
    return provider.lazily(clazz, expression, null);
  }

  @SuppressWarnings("unchecked")
  public <S> IRavenQueryable<S> as(Class<S> resultClass) {
    this.clazz = (Class<T>) resultClass;
    this.provider = provider.forClass(resultClass);
    return (IRavenQueryable<S>) this;
  }

  @Override
  public IRavenQueryable<T> orderBy(OrderSpecifier< ? >... asc) {
    return getProvider().createQuery(Expressions.operation(expression.getType(), LinqOps.Query.ORDER_BY, expression, Expressions.constant(asc)));
  }


}
